http/etag.ts

```diff
@@ -0,0 +1,196 @@
+ * ```
+ */
+export async function calculate(
+  entity: Entity,
+  options: ETagOptions = {},
+): Promise<string | undefined> {
+  const weak = options.weak ?? isFileInfo(entity);
+  const tag =
+    await (isFileInfo(entity)
+      ? calcFileInfo(entity, options)
+      : calcEntity(entity, options));
+
+  return tag ? weak ? `W/"${tag}"` : `"${tag}"` : undefined;
+}
+
+/** A helper function that takes the value from the `If-Match` header and a
+ * calculated etag for the target. By using strong comparison, return `true` if
+ * the values match, otherwise `false`.
+ *
+ * See MDN's [`If-Match`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match)
+ * article for more information on how to use this function.
+ *
+ * ```ts
+ * import {
+ *   calculate,
+ *   ifMatch,
+ * } from "https://deno.land/std@$STD_VERSION/http/etag.ts";
+ * import { serve } from "https://deno.land/std@$STD_VERSION/http/server.ts";
+ * import { assert } from "https://deno.land/std@$STD_VERSION/_util/asserts.ts"
+ *
+ * const body = "hello deno!";
+ *
+ * await serve(async (req) => {
+ *   const ifMatchValue = req.headers.get("if-match");
+ *   const etag = await calculate(body);
+ *   assert(etag);
+ *   if (!ifMatchValue || ifMatch(ifMatchValue, etag)) {
+ *     return new Response(body, { status: 200, headers: { etag } });
+ *   } else {
+ *     return new Response(null, { status: 412, statusText: "Precondition Failed"});
+ *   }
+ * });
+ * ```
+ */
+export function ifMatch(
+  value: string | null,
+  etag: string | undefined,
+): boolean {
+  // Weak tags cannot be matched and return false.
+  if (!value || !etag || etag.startsWith("W/")) {
+    return false;
+  }
+  if (value.trim() === "*") {
+    return true;
+  }
+  const tags = value.split(/\s*,\s*/);
+  return tags.includes(etag);
+}
+
+/** A helper function that takes the value from the `If-None-Match` header and
+ * a calculated etag for the target entity and returns `false` if the etag for
+ * the entity matches the supplied value, otherwise `true`.
+ *
+ * See MDN's [`If-None-Match`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match)
+ * article for more information on how to use this function.
+ *
+ * ```ts
+ * import {
+ *   calculate,
+ *   ifNoneMatch,
+ * } from "https://deno.land/std@$STD_VERSION/http/etag.ts";
+ * import { serve } from "https://deno.land/std@$STD_VERSION/http/server.ts";
+ * import { assert } from "https://deno.land/std@$STD_VERSION/_util/asserts.ts"
+ *
+ * const body = "hello deno!";
+ *
+ * await serve(async (req) => {
+ *   const ifNoneMatchValue = req.headers.get("if-none-match");
+ *   const etag = await calculate(body);
+ *   assert(etag);
+ *   if (!ifNoneMatch(ifNoneMatchValue, etag)) {
+ *     return new Response(null, { status: 304, headers: { etag } });
+ *   } else {
+ *     return new Response(body, { status: 200, headers: { etag } });
+ *   }
+ * });
+ * ```
+ */
+export function ifNoneMatch(
```

Comment:
Is this really needed? Couldn't this be done with `!ifMatch()`?

Reply:
No the logic is quite different. If-None-Match uses weak matching, If-Match doesn't.